storage/range.go

```diff
@@ -606,6 +606,24 @@ func (r *Range) addReadWriteCmd(args proto.Request, reply proto.Response, wait b
 		// occurred after our txn timestamp.
 		if !rTS.Less(header.Timestamp) {
 			header.Timestamp = rTS.Next()
+
+			// Check if a SSI transaction's timestamp is pushed by rts,
```

Comment:
The goal with the current implementation is to lay down write intents to make the retry more likely to succeed. Every write intent which is persisted prevents subsequent reads from succeeding (without pushing this txn). This is important, otherwise a busy key for reads will always prevent an SSI txn from completing. It will continually restart, come back to read the key, but discover again that the timestamp has been moved forward.

Going through the entire txn and laying down write intents is slower if you only expect a single retry, but it prevents starvation. Also, OLTP transactions are typically quite short, so I think it's better to just let it run through until the EndTransaction.